docs/source/federation-spec.mdx

```diff
@@ -0,0 +1,625 @@
+#   must include every object type in the schema that uses
+#   the @key directive (i.e., all federated entities).
+union _Entity
+
+scalar _Any
+scalar FieldSet
+scalar link__Import
+
+enum link__Purpose {
+  """
+  `SECURITY` features provide metadata necessary to securely resolve fields.
+  """
+  SECURITY
+
+  """
+  `EXECUTION` features provide metadata necessary for operation execution.
+  """
+  EXECUTION
+}
+
+type _Service {
+  sdl: String
+}
+
+extend type Query {
+  _entities(representations: [_Any!]!): [_Entity]!
+  _service: _Service!
+}
+
+directive @external on FIELD_DEFINITION
+directive @requires(fields: FieldSet!) on FIELD_DEFINITION
+directive @provides(fields: FieldSet!) on FIELD_DEFINITION
+directive @key(fields: FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
+directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA
+directive @shareable on OBJECT | FIELD_DEFINITION
+directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
+directive @override(from: String!) on FIELD_DEFINITION
+
+# This is an optional directive (see the glossary below)
+directive @extends on OBJECT | INTERFACE
+```
+
+## Enhanced introspection with `Query._service`
+
+A federated graph router (also known as a gateway) can compose its supergraph schema dynamically at runtime. To do so, it first executes the following enhanced introspection query on each of its subgraphs to obtain all subgraph schemas:
+
+```graphql
+query {
+  _service {
+    sdl
+  }
+}
+```
+
+> We strongly recommend _against_ performing composition dynamically within the graph router, because a composition error can cause unexpected downtime. Nevertheless, supporting this use case is still a requirement for subgraph libraries.
+
+### Differences from built-in introspection
+
+The "enhanced" introspection query above differs from [the GraphQL spec's built-in introspection query](https://spec.graphql.org/October2021/#sec-Schema-Introspection) in the following ways:
+
+- The returned schema representation is a string instead of a `__Schema` object.
+- The returned schema string includes all uses of federation-specific directives, such as `@key`.
+    - The built-in introspection query's response does _not_ include the uses of any directives.
+    - The graph router requires these federation-specific directives to perform composition successfully.
+-  If a subgraph server "disables introspection", the enhanced introspection query is still available.
+
+> Note that the `_service` field is _not_ included in the composed supergraph schema. For security reasons, it's intended solely for use by the graph router.
+
+### Required resolvers
+
+To support the enhanced introspection query, a subgraph service must define resolvers for the following fields:
+
+- `Query._service`
```

Comment:
instead maybe provide this info as GraphQL SDL (as otherwise you will need to specify the nullability)?

```
extend type Query {
  _service: _Service!
}

type _Service {
  sdl: String!
}
```

Reply:
technically the only required resolver would be for `_entities` query if there is a federated type (type with `@key`) in the schema

Reply:
Good idea!